README.md
# NVRAM Save/Restore Utility for Asuswrt-Merlin Firmware
[![Build Status](https://travis-ci.com/Xentrk/nvram-save-restore-utility.svg?branch=master)](https://travis-ci.com/Xentrk/nvram-save-restore-utility)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/dd74bea80e1846ddbbde7d794aff2b73)](https://www.codacy.com/manual/Xentrk/nvram-save-restore-utility?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=Xentrk/nvram-save-restore-utility&amp;utm_campaign=Badge_Grade)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

## Introduction
Utility to save/restore NVRAM values and the jffs partition on Asuswrt-Merlin firmware.

## Support
For help and support, please visit the NVRAM Save/Restore Utility support thread on [snbforums.com](https://www.snbforums.com/threads/release-nvram-save-restore-utility.61722/).

See the
[[FAQ] NVRAM and Factory Default Reset](https://www.snbforums.com/threads/faq-nvram-and-factory-default-reset.22822/) for additional information.

## Requirements
A properly formatted USB drive with an available partition for the NVRAM Save/Restore Utility installation.

## Project Development
[john9547](https://www.snbforums.com/members/john9527.27638/) is the original author of the NVRAM Save/Restore Utility. The last update made by john9527 was [Version 26.2](http://bit.ly/2aaAySO) released on 24-Sep-2018.

[Martineau](https://www.snbforums.com/members/martineau.13215/) made numerous updates to the NVRAM Save/Restore Utility:
* Version 30.1 code updates to **nvram-restore.sh**, **nvram-restore.sh** and **nvram-save.sh** made the utility compatible with the 384.13 Asuswrt-Merlin release.
* Added functionality to **nvram-restore.sh** listing the restore files with the ability to delete a restore file or a range of restore files.
* Add functionality to **jffs-restore.sh** listing the jffs restore files.

[Xentrk](https://www.snbforums.com/members/xentrk.49161/) contributed the hosting of the repository on GitHub, coding of the installation menu, code updates to make the code POSIX compliant, and updating the nvram variables in the nvram-merlin.ini file.

Contributions to the project by other coders are always welcome and appreciated.

## Installation

Copy and paste the command below into an SSH session:

```
/usr/sbin/curl --retry 3 "https://raw.githubusercontent.com/Xentrk/nvram-save-restore-utility/master/nsrum" -o "/jffs/scripts/nsrum" && sleep 5 && chmod 755 /jffs/scripts/nsrum && sh /jffs/scripts/nsrum
```
This command will download and install the NVRAM Save/Restore Ulitity Menu **nsrum** to the **/jffs/scripts** directory.

<img src="https://github.com/Xentrk/nvram-save-restore-utility/blob/master/InstallationMenu.png" alt="drawing" width="700" height="600"/>

Option **[5]  Update NVRAM Save/Restore Utility Menu** will only be displayed if there is an update to the installation menu available.

Select option **[1]  Install NVRAM Save/Restore Utility** to install the utility. You will be presented with a list of available partitions on the USB drive. Select the partition you want to install the utility in. A directory called **nsru** (the acronym for NVRAM Save/Restore Utility) will be created in the partition for the utility scripts and files (e.g. **/mnt/ASUS/nsru**). The installation program will also create a backup directory (e.g. **/mnt/ASUS/nsru/backup**) to keep the generated save/restore files separate from the utility scripts and files.

To access the installation menu, type the command **nsrum** at the command prompt. To access the project directory, type the command **nsru** at the command prompt. These commands will not work after performing a factory reset since they require alias entries in **/jffs/configs/profile.add**, which will not exist after performing a factory reset. After a factory reset, you must navigate to the NVRAM Save/Restore Utility directory using the **cd** command (e.g. cd /mnt/AC88U/nsru).

## Project Scripts and Files
| Script Name   | Description |
| --- | --- |
| nvram-save.sh |	Main user script |
nvram-restore.sh | User script to restore saved user NVRAM variables |
| jffs-restore.sh	|	User script to restore jffs directory |
| nvram-hnd-migr.sh | Script to restore nvram parameters when migrating from a non-HND router to a HND* router |
| nvram-excp-merlin.sh | Script to adjust nvram parameters for code specific issues |
| nvram-merlin.ini | NVRAM variable control file
| nvram-sample.ini | Example user NVRAM control file with only basic variables |
| nvram-user-sample.sh | Creates a tar backup of the nvram backup files and jffs directory |
| clear-maclist.sh	| Script to clear MAC list filters in case of problems detecting the correct format (multiple changes between firmware levels) |
| save-mynvram.sh	|	A small script that can save/restore a small number of user specified nvram settings |
| QuickStart.txt	|	Documentation in text format |
| Changelog.txt |		Version history in text format |

*HND Router models: RT-AC86U and RT-AX88U

## Running the Scripts from the Command Line
The NVRAM Save/Restore Utility scripts must be run from the command line in the project directory.
* At the SSH prompt, enter ```ls /mnt``` to see the names of the partition labels on the USB stick (e.g. ASUS).
* Change to the project directory using the change directory command (e.g. ```cd /mnt/ASUS/nsru```).
* Run the scripts using the syntax ```sh script-name.sh``` or ```./script-name.sh```.

## nvram-save.sh Usage Notes

* Basic help is available by entering ```sh nvram-save.sh -h```

```
NVRAM User Save/Restore Utility
nvram-save.sh Version 30.3.0
Options: -h           this help msg
         -v           Print version/perform consistency check
         -b           Backup mode - save for restore to same router (default)
         -m           Migration mode - transfer settings to another router
         -i inifile   Specify custom nvram variable ini file
         -clk         Include clkfreq/overclock setting (Backup mode only)
         -nojffs      Skip backup of jffs storage
         -nouser      Skip execution of user exit script
```
* When running the utility or using the -v version option, **nvram-save.sh**, **nvram-restore.sh**, **jffs-restore.sh**, **nvram-merlin.ini** and **nvram-excp-merlin.sh** are consistency checked to make sure they are all the same version. A WARNING message will be printed on a version mismatch, but execution will continue if actually running to generate the restore script.

* You can now specify a custom nvram variable 'ini' file to use to generate the restore script. A sample ini file, **nvram-sample.ini**, is included which is a stripped down version of the full ini containing only basic entries that people have mentioned in various threads.

* The default state is NOT to save clkfreq overclocking values (safest when moving to a new code level). If you wish to include this setting, specify the -clk option on the nvram-save.sh command.

* The nvram-merlin.ini files contains a section at the end for additions. You may edit this section to add any NVRAM variables you may have created for use in scripts, or to temporarily add any variables which may be missing.

* The NVRAM Save/Restore Utility also creates three text files during execution that can be viewed or archived.
  * nvram-all-yyyymmddhhmm_MODEL-macid.txt (all nvram variables, including system variables not normally changed by the user)
  * nvram-usr-yyyymmddhhmm_MODEL-macid.txt (the nvram variables actually saved by the utility in their save categories)
  * nvram-ini-yyyymmddhhmm_MODEL-macid.txt (a copy of the nvram-merlin.ini configuration file used by the utility in the nvram-save operation)

  Where
  * yyyymmddhhmm is the timestamp of the saved setting
  * MODEL is the model number e.g. RT-AC88U
  * macid In Backup Mode, this is the last two bytes of the router MAC address (for example E1F2). In Migration Mode, this will be MIGR

* The utility is still valid for ASUS OEM as well as Merlin releases, although it will not attempt to fix changes in the use of nvram variables between releases (just too many ASUS levels with various release numbers to track).

## nvram-restore.sh Usage Notes

* Enter ```sh nvram-restore.sh```

You will be presented with a listing of current nvram-restore backup files. The **'Ver=--------'** indicates the presence of a nvram-restore file that was interrupted when nvram-save.sh was run.

Enter a 'y' or 'Y' to restore nvram using the most recent file. Press the enter key or any non-valid character to exit.

To delete a NVRAM restore file (and all associated text files) and the associated jffs restore backup directory, enter the line number followed by the 'del' parameter (e.g. 1 del). To delete a range of files, enter the from-to range separated by a dash followed by the 'del' parameter (e.g. 1-3 del).

```
          nvram-restore.sh NVRAM Restore Utility - Version 30.3.0

        NVRAM Restore File Directory: /tmp/mnt/AC88U/nsru/backup

  1 : nvram-restore-202001112027_RT-AC88U-8248.sh       Ver=384.14_2 nvram-merlin.ini
  2 : nvram-restore-202001112031_RT-AC88U-8248.sh       Ver=384.14_2 nvram-merlin.ini
  3 : nvram-restore-202001121854_RT-AC88U-8248.sh       Ver=384.14_2 nvram-merlin.ini
  4 : nvram-restore-202001151944_RT-AC88U-8248.sh       Ver=--------
  5 : nvram-restore-202001151945_RT-AC88U-8248.sh       Ver=--------

        Restore last NVRAM save ? (nvram-restore-202001121854_RT-AC88U-8248.sh)
        { n[{-n}] [ del ] | Y } (or press ENTER to ABORT) > 5 del
        Deleted                   (*202001151945_RT-AC88U-8248.*)

```
You can also provide the yyyymmddhhmm_MODEL-macid code on the command line to load a specific file:

```
sh nvram-restore.sh 202001121854_RT-AC88U-8248

Migration restore from a full nvram-save is not supported on this save file

Do you want to proceed [Y/N]? y

A clean restore will restore only those NVRAM settings
  which have been initialized by the router after a factory reset
This option should by used when reverting to an earlier
  firmware level, when updating past a single firmware release,
  when moving between OEM and Merlin firmware or
  to ensure any special use or obsolete NVRAM variables are cleared
If in doubt, answer Y to perform a clean restore

Perform a clean restore and remove unused NVRAM variables [Y/N]?
```
### Restoring After a Factory Reset

* The **UPnP** setting on the WAN page must be configured manually after running **nvram-restore.sh**.
* You will need to acknowledge the **TrendMicro** EULA upon logging into the router Web GUI if it was previously enabled.

## jffs-restore.sh Usage Notes

* Enter ```sh jffs-restore.sh```

You will be presented with a listing of current jffs backup files.

Enter a 'y' or 'Y' to restore the jffs partition, or the enter key, or any non-valid character, to exit.

```
  jffs-restore.sh JFFS Restore Utility - Version 30.3.0

  JFFS Restore File Directory: /tmp/mnt/AC88U/nsru/backup

     1 : jffs-202001112027_RT-AC88U-8248
     2 : jffs-202001112031_RT-AC88U-8248
     3 : jffs-202001121854_RT-AC88U-8248

  Restore last saved /jffs/ ? (jffs-202001121854_RT-AC88U-8248)
  { n | Y } (or press ENTER to ABORT) >
```

You can also provide the yyyymmddhhmm_MODEL-macid code on the command line to load a specific file:

```
sh jffs-restore.sh 202001121854_RT-AC88U-8248

        Restoring saved /jffs/ (jffs-202001121854_RT-AC88U-8248)

        WARNING: This will overwrite/replace the current contents of your /jffs directory!
        Do you want to continue [Y/N] ?
```
## nvram-merlin.ini File Format and Usage Notes
**nvram-merlin.ini** is the NVRAM variable control file used by **nvram-save.sh** containing the nvram variables to process. The **nvram-merlin.ini** file format for the nvram definitions follows the format of a section header followed by nvram variables. For example:
```
[System - Basic]
time_zone_dst
time_zone
time_zone_dstoff
```
There are additional controls which can be added. To exclude an entire section from backup, comment # the section header
```
#[System - Basic]
time_zone_dst
time_zone
time_zone_dstoff
```

To exclude an individual nvram setting from backup, comment # the setting
```
[System - Basic]
#time_zone_dst
time_zone
time_zone_dstoff
```
To exclude the setting from being saved during a migration save, or restored during a migration restore, prefix it with a @. This is only valid for individual settings and is not valid on the section header.
```
[System - Basic]
@time_zone_dst
time_zone
time_zone_dstoff
```
To force a setting to always be included in the restore, even if it would normally be excluded on a 'clean' restore, prefix the setting with a + .  This may also be used to preserve custom nvram settings across a factory reset. This is only valid for individual settings and is not valid on the section header.
```
[System - Basic]
+time_zone_dst
time_zone
time_zone_dstoff
```

To force a setting to always be included in the save, even if it is an empty value, prefix the setting with a % .  This will force some variables which may have a valid empty setting to be saved/restored. This is only valid for individual settings and is not valid on the section header.
```
[System - Basic]
%time_zone_dst
time_zone
time_zone_dstoff
```

To combine the above flags and force both the save of an empty value and force it to be restored even if it is not present after a factory reset, prefix the setting with a & .
This is only valid for individual settings and is not valid on the section header.
```
[System - Basic]
&time_zone_dst
time_zone
time_zone_dstoff
```

The last section header in the default nvram-merlin.ini file is for user added nvram settings, such as may be used in custom scripting. For example (also using the force save prefix) for a user nvram variable 'blacklist_enable'

```
[User Adds]
+blacklist_enable
```

## save-mynvram.sh Usage Notes


Changelog.txt
Changelog
---------
Version 30.4	  18-November-2020
Program Changes
- update nvram-merlin.ini to include john9547 fork nvram parameters

Version 30.3	  10-February-2020
Program Changes
- POSIX and code updates for launch to SNBForums.

Version 30.1
Program Changes
- Version 30.1 Martineau Hacked for v384.xx+!!

Version 26.2	  22-September-2017
Program Changes
- fix problem honoring special variable flags during restore (nvram-save.sh change)


Version 26.1	  15-September-2017
Program Changes
- update prompts during jffs-restore to better address migrations
- process tagged vars that need to have empty values saved/restored
- add support for separate user variable adds via nvram.ini.add
INI file changes
- tag dsn_probe and dns-filter vars for forced save/restore even if empty
- resync with new adds for 380.68


Version 26.0	  1-September-2017
Program Changes
- change version numbering
- updates for migration and clean restore
- alert for mac address mismatch instead of aborting
- change for QoS inactive after restore
- add vpn client accept dns changes to exception processing
INI file Changes
- change version numbering
- add missing enable vars for ftp and samba
- force restore transient var qos_inuse
- exclude samba and media server names during migrate (default based on mac)
- sync with 380.68 and V27 final


Version 25a	  9-August-2017
Program Changes
- change force include flag to '+' to work around busybox bug
INI file Changes
- change force include flag to '+' to work around busybox bug


Version 25	  8-August-2017
Program Changes
- support ini file ignore during clean variable flag
- support doing a migration restore from a full save file
- fix run log formatting in nvram-restore and jffs-restore
- refactor exception processing and log changes made
- update exception processing to handle sshd options
- misc QuickStart guide updates
INI file Changes
- sync with Merlin 380.67 and LTS fork V27
- misc updates to wan and wifi variables
- add support for hotspot and captive portal features (currently not supported)


Version 24a	  12-February-2017
INI file Changes
- sync with 380.65 Merlin and V23 fork
- add support for Merlin firmware update
- add support for new ipv6 options
- add support for OpenVPN 2.4
- add support for DNSCrypt
- add support MU-MIMO options
- add support for Bluetooth compatibility
- add support for Facebook WIFI (ASUS OEM only)
- misc updates to wan and wifi variables


Version 24	  2-May-2016
Program Changes
- add hhmm to file timestamp naming (full timestamp)
- set qos_type if moving to adaptive qos firmware from non-adaptive
- update Quickstart guide with info on need for jffs restore
- update Quickstart guide with info on moving MAC list filters
  between firmware levels
INI file Changes
- fix 3rd radio psk and wps vars
- add Smartconnect support


Version 23	  3-April-2016
Program Changes
- handle mac_list differences between fw levels
- include clear-maclist-sh to reset MAC filters if necessary
- auto correct bad wl_band vars (fix old bug still present on some systems)
- some changes dependent on not in migration mode
- remove code to delete unused vars (use clean restore instead)
- exceptions not applied if ASUS OEM code detected
- ensure do not enable scripts in migration mode
- add help text for clean restore
- update Quickstart guide with Apple OS info
- single Quicsktart.txt guide in DOS/Windows format
INI file Changes
- sync with 380.58 Merlin and V17 fork
- reduced number of OpenVPN servers to 2 (number actually supported)
- move and update vars to new Tweaks and Hidden settings section
- add support for radius server
- add user Email vars
- add custom icon vars
- add Traffic Control settings (future fw)
- added section for custom User Adds


Version 22	  27-October-2015
Program Changes
- show version numbers with -v
- print ini version to syslog
- add ini version to runlog
- only set file permissions on non-FAT format drives
- update Merlin/Fork unique variables
INI file Changes
- fix SMB/FTP access restore by adding index var
- fix ipv6 settings for non-native modes
- additional support for 3 radios
- add support for 374 fork http port change
- add support for syslog logging level and msg level
- add support for reboot schedule
- add support for up to 5 VPN servers/clients
- add support for QOS bw limiting
- add support for WTFast
- add support for KeyGuard


Version 21	  6-August-2015
Program Changes
- fix call to exception processing if not running from current directory
- fix bug in setting jffs options moving from <378 to 378.55


Version 20        3-August-2015
Program Changes
- enable running script from other than current directory
- strip CR from custom strings during save
- update Merlin/Fork unique vars in exception processing
INI file Changes
- add missing Protect Management Frame vars
- add missing mode Settings vars
- add missing Disk Utility vars
- add support for AC3200 third radio
- add support for Concurrent Repeater (future fw)
- add support for bandwidth monitoring (future fw)
- update Merlin/Fork unique vars


Version 19         16-June-2015
INI file Changes
- restore overzealous cleaning of dhpc_, lan_ vars
- add new Merlin jffs2_enable


Version 18         14-June-2015
Program Changes
- add -clean option (for nvram-restore) to only
  only restore nvram variables which exist
  (useful if backleveling fw to clean
  nvram variables added for new function)
- running in migration mode automatically invokes
  the -clean option when restoring
- nvram-restore prompts for clean restore with file generated
  by latest level
- updated exception script with latest fork variables
- update QuickStart guide
INI file Changes
- general update based on latest defaults.c
  included reported missing
  ipv6_ns_drop, dms_dir_manual, acs_dfs
- Add [TOR VPN] section
- remove unneccessary wl_,dhcp_ vars
- update vars for migration mode


Version 17        29-April-2015
INI file changes only
- added missing var for http/https mode

nvram-restore.sh
#!/bin/sh
# shellcheck disable=SC1001 # SC1001: This \/ will be a regular '/' in this context.
# shellcheck disable=SC2012 # Ignore Use find instead of ls to better handle non-alphanumeric filenames. Vers of find needed is only available in entware?
# shellcheck disable=SC2034 # Ignore "color appears unused" msgs
# shellcheck disable=SC2068 # Don't want to quote array $@
###########################################################################################################################################
#Version=30.4.0
VERSION=30.4.0
# nvram-restore.sh
# General front end to calling generated nvram-restore-yyyymmdd-macid scripts
# Supports OEM and Merlin firmware ONLY
#
# Changelog
#------------------------------------------------
# Version 30.4  18-November
# - Include nvram variables from John's Fork
#
# Version 30.3			  12-January-2020
# - Xentrk POSIX Updates
#
# Version 30.1			    xx-xxxxxxxxx-20xx
# - Martineau Hacks
# - Process list of files with options to delete one or a range of files
# see nvram-save.sh    	nvram-save.sh: WARNING: nvram-merlin.ini is mismatched version
#												nvram-save.sh: WARNING: nvram-excp-merlin.sh is mismatched version
#												nvram-save.sh: WARNING: jffs-restore.sh is mismatched version
#
# Version 26.1			    15-September-2017
# - minor prompt change
#
# Version 26.0			    16-Aug-2017
# - change version numbering
# - updates for migration and clean restore
# - alert for mac address mismatch instead of aborting
#
# Version 25			    8-Aug-2017
# - support doing a migration restore from a full save file
# - fix run log formatting
#
# Version 24			    2-May-2016
# - version update only
#
# Version 23			    3-April-2016
# - add help text for clean restore
#
# Version 22			    27-October-2015
# - add ini version to runlog
# - version variable name update
#
# Version 21			    6-August-2015
# - version update only
#
# Version 20			    3-August-2015
# - enable running script from other than current directory
#
# Version 19			     16-June-2015
# - version update only for ini
#
# Version 18			     14-June-2015
# - prompt for clean restore with file generated
#   by latest level
# - add codelevel to runlog
# - added exit for no valid restore scripts
#
# Version 17			    29-April-2015
# - version update only for ini
#
# Version 16			    16-April-2015
# - first release as part of Version 16 package
# - use last run for restore if none specified
#
############################################################################################EIC Hack ######################################

ANSIColours() {

  cRESET="\e[0m"
  cBLA="\e[30m"
  cRED="\e[31m"
  cGRE="\e[32m"
  cYEL="\e[33m"
  cBLU="\e[34m"
  cMAG="\e[35m"
  cCYA="\e[36m"
  cGRA="\e[37m"
  cBGRA="\e[90m"
  cBRED="\e[91m"
  cBGRE="\e[92m"
  cBYEL="\e[93m"
  cBBLU="\e[94m"
  cBMAG="\e[95m"
  cBCYA="\e[96m"
  cBWHT="\e[97m"
  aBOLD="\e[1m"
  aDIM="\e[2m"
  aUNDER="\e[4m"
  aBLINK="\e[5m"
  aREVERSE="\e[7m"
  cRED_="\e[41m"
  cGRE_="\e[42m"
}

Parse() {

  # 	Parse		"Word1,Word2|Word3" ",|" VAR1 VAR2 REST
  #				(Effectively executes VAR1="Word1";VAR2="Word2";REST="Word3")
  OLD_IFS=$IFS
  TEXT="$1"
  IFS="$2"
  shift 2
  read -r $@ <<EOF
$TEXT
EOF
  IFS=$OLD_IFS
}

Validate_Input() {

  INPUT=$1

  Parse "$INPUT" " " INPUT ACTION
  echo "$INPUT" | grep -q "\-" && Parse "$INPUT" "- " INPUT MATCH_TO
  LISTNUM=1
  #MULTI=
  MATCH_FROM=$INPUT
  MATCH="9999"
  #if [ -n "$MATCH_TO" ] && [ -z "$(echo $MATCH_TO | grep -E "^[1-9]+$")" ];then   # Prevent accidental say 6-99 deletion when only 6-9 intended!
  ##if [ -n "$MATCH_TO" ] && [ -z "$(echo "$MATCH_TO" | grep -E "^[1-9]+[0-9]*$")" ]; then # Remove safeguard
  if [ -n "$MATCH_TO" ] && echo "$MATCH_TO" | grep -qvE "^[1-9]+[0-9]*$"; then # Remove safeguard
    MATCH_TO=$MATCH_FROM
    printf '%b\a\t%s\n\n%b' "$cBRED" "***ERROR valid number selection/range" "$cRESET" && \
    exit 96
  fi

  if [ -z "${INPUT##*[!0-9]*}" ]; then # Try and trap "8^H" where backspace is wrong!
    printf '%b\a\n\t%s\n\n%b' "$cBRED" "***ERROR Invalid number $INPUT" "$cRESET"
    exit 97
  else
    if [ "$INPUT" -gt "$(echo "$DIR_LISTING" | wc -w)" ]; then
      printf '%b\a\n\t%s%c%s%c%s\n\n%b' "$cBRED" "***ERROR selection " "'" "$INPUT" "'" " out of range" "$cRESET"
      exit 98
    fi
  fi

  if [ -z "$MATCH_TO" ]; then
    MATCH_TO=$MATCH_FROM
  fi

  for ITEM in $DIR_LISTING; do
    if [ "$ACTION" = "del" ] && [ "$LISTNUM" -ge "$MATCH_FROM" ] && [ "$LISTNUM" -le "$MATCH_TO" ]; then
      MATCH=$LISTNUM
    fi
    #echo "DEBUG From=$MATCH_FROM To=$MATCH_TO Match=$MATCH Item=$ITEM Listnum=$LISTNUM"
    if [ "$LISTNUM" -eq "$MATCH" ]; then
      restore_version=$(echo "$ITEM" | sed 's/.sh//')                       # Strip '.sh' suffix
      restore_version=$(echo "$restore_version" | sed 's/nvram-restore-//') # Strip 'nvram-restore-' prefix
      restore_version=$(echo "$restore_version" | sed "s|$dwd||;s|/||")     # strip out directory path and leading /
      if [ -z "$ACTION" ]; then
        break
      else
        if [ "$ACTION" = "del" ] || [ "$ACTION" = "delX" ]; then
          if [ "$LISTNUM" -eq "$MATCH" ] && [ "$(echo $@ | grep -c "test")" -eq 0 ]; then
            if [ "$TEST" != "testmode" ]; then
              rm "${dwd}/nvram-restore-${restore_version}.sh" 2>/dev/null
              rm "${dwd}/nvram-all-${restore_version}.txt" 2>/dev/null
              rm "${dwd}/nvram-usr-${restore_version}.txt" 2>/dev/null
              rm "${dwd}/nvram-ini-${restore_version}.txt" 2>/dev/null
              rm "${dwd}/nvram-backup-${restore_version}.tar.gz" 2>/dev/null
              # ...and the JFFS directory
              [ -d "${dwd}/jffs-${restore_version}" ] && rm -rf "${dwd}/jffs-${restore_version}"
              # ...and the log entry?
              #grep -v "${restore_version}" nvram-util.log > nvram-util.log2; mv nvram-util.log2 nvram-util.log
              sed -i "/${restore_version}/d" nvram-util.log
              printf '%b\t%s%b%s%b%s\n\n' "$cBWHT$aBOLD" "Deleted                   (" "$cRED" "*$restore_version.*" "$cBWHT" ") "
            else
              printf '%b\t%s%b%s%b%s\n\n' "$cBWHT" "Test: Deleted                   (" "$cRED" "*$restore_version.*" "$cBWHT" ") "
            fi
          fi
        else
          printf '%b\a\n\t%s%c%s%c\n\n%b' "$cBRED" "***ERROR action " "'" "$ACTION" "'" "$cRESET"
          break
        fi
      fi
    fi
    LISTNUM=$((LISTNUM + 1))
  done
  printf '%b' "$cRESET"
}

ANSIColours
#MYROUTER=$(nvram get computer_name)
#MYROUTER="$(nvram get model)"
if [ "$1" = "test" ]; then
  TEST="testmode"
  shift
fi
########################################################################################################################
restore_version=$1
opt=""
scr_name=$(basename "$0")
#cwd=$(pwd)
cwd="$(dirname "$(readlink -f "$0")")"
if [ -d "$cwd/backup" ]; then
  dwd=$cwd/backup
else
  dwd=$cwd
fi
runlog="$cwd/nvram-util.log"
space4="    "
codelevel=""

echo ""
if [ "$restore_version" = "" ] || [ "$1" = "Y" ]; then
  macid=$(nvram get "et0macaddr")
  #  if [ "$macid" = "" ]; then
  if [ "$macid" = "" ] || [ "$macid" = "00:00:00:00:00:00" ]; then
    macid=$(nvram get "et1macaddr")
    if [ "$macid" = "" ] || [ "$macid" = "00:00:00:00:00:00" ]; then
      # Xentrk noticed MACID is all 0"
      macid=$(nvram get "et2macaddr")
    fi
  fi
  macid=$(echo "$macid" | cut -d':' -f 5,6 | tr -d ':' | tr '[:upper:]' '[:lower:]')
  restore_version=$(grep "nvram-save.sh" "$runlog" | tail -1 | awk -F' ' '{print $2}')
  if [ "$restore_version" = "" ]; then
    echo "No valid nvram-save file found. Exiting."
    echo ""
    exit 3
  fi
  macsave=$(printf '%s' "$restore_version" | tail -c 4)

  if [ "$1" != "Y" ]; then
    if [ "$macsave" != "$macid" ]; then
      #################################################################################EIC hack###############################
      printf '%b' "$cRED"
      ########################################################################################################################
      echo "Last nvram save/restore (nvram-restore-$restore_version.sh)"
      echo "WARNING: Last nvram-save MAC address ($macsave)"
      echo "  does not match current router MAC  ($macid)!"
      echo "This may be valid if you are migrating settings from one router to another"
      echo ""
      echo "If you wish to use a different restore file, answer N and specify the"
      echo "  restore file id the command line in the form"
      echo "  yyyymmddhhmm-mmmm"
      #################################################################################EIC hack###############################
      printf '%b' "$cRESET"
      printf '%s' "Do you want to proceed [Y/N]? "
      read -r "INPUT"
      while true; do
        case "$INPUT" in
        "Y" | "y")
          break
          ;;
        *)
          printf '%b\n\t%s\n\n%b' "$cBYEL" "Restore aborted" "$cRESET"
          exit 2
          ;;
        esac
      done
      echo ""
    fi
    #################################################################################EIC hack###############################
    I=1
    # Only list files that were created by nvram-save (in this millenium! 20xx)
    #	i.e. ignore non standard files e.g. 'nvram-restore-OVPN.sh' etc.

    DIR_LISTING="$(ls -F -- "$dwd"/*nvram-restore-20* | sed 's/*//' 2>/dev/null)"
    FILE_NAME=$(echo "$DIR_LISTING" | sed 's|'"${dwd}"\/'||')
    #DIR_LISTING=$(/usr/bin/find . -maxdepth 1 -type f | sort | grep "nvram-restore-20" | sed 's/.\///')
    printf '\t%s\n\n' "  $scr_name NVRAM Restore Utility - Version $VERSION"
    printf '\t%b%s%b%s\n\n' "$cBBLU" "NVRAM Restore File Directory: " "$cBWHT" "$dwd"
    for ITEM in $FILE_NAME; do
      #TIMESTAMP=$(echo "$ITEM" | awk '{ string=substr($0, 37, 12); print string; }')
      TIMESTAMP=$(echo "$ITEM" | awk '{ string=substr($0, 15, 12); print string; }')
      VER=$(grep -m 1 "$TIMESTAMP" nvram-util.log | awk '{print $9" "$10}' | sed 's/ $//') # Martineau FIX Xentrk
      if [ -z "$VER" ]; then
        VER="--------"
      fi
      printf '%b%3d : %b%s%b\t%s\n' "$cBWHT" "$I" "$cBBLU" "$ITEM" "$cBCYA" "Ver=$VER"
      I=$((I + 1))
    done

    printf '%b\n\t%s%b%s%b%s\n' "$cBWHT" "Restore last NVRAM save ? (" "$aBOLD$cBYEL" "nvram-restore-$restore_version.sh" "$cBWHT" ") "
    printf '\t%s' "{ n[{-n}] [ del ] | Y } (or press ENTER to ABORT) > "
    read -r "INPUT"
    printf '%b' "$cRESET"
    ########################################################################################################################
    while true; do
      case "$INPUT" in
      "Y" | "y")
        break
        ;;
        #################################################################################EIC hack###############################
        #*)
      "N" | "n")
        printf '%b\n\t%s\n\n%b' "$cBYEL" "Restore aborted" "$cRESET"
        exit 1
        ;;
        #################################################################################EIC hack###############################
      *)
        echo "$INPUT" | grep -qE "^[1-9]" && Validate_Input "$INPUT" || printf '%b\n\t%s\n\n%b' "$cBYEL" "Restore aborted" "$cRESET" && exit 0
        if [ -z "$ACTION" ]; then
          printf '%b\t%s%b%s%b%s' "$cBWHT" "Restoring    NVRAM save (" "$aBOLD$cBGRE" "nvram-restore-$restore_version.sh" "$cBWHT" ")"
        else
          exit 0
        fi
        ;;
        ########################################################################################################################
      esac
    done
  fi
fi
restore_file="nvram-restore-$restore_version.sh"
echo ""
if [ ! -f "$dwd/$restore_file" ]; then
  echo "Script $dwd/$restore_file does not exist. Exiting."
  echo ""
  exit 1
fi

if [ "$1" = "Y" ]; then
  opt="-clean"
fi

if grep -q "Migrate restore from full restore supported" "$dwd"/"$restore_file" && [ "$macsave" != "MIGR" ]; then
  if [ "$opt" = "" ] && [ "$macsave" != "$macid" ]; then
    echo ""
    #################################################################################EIC hack###############################
    printf '%b' "$cBYEL"
    echo "A migration restore is used when transferring settings from one"
    echo "  router to another, even if they are both the same model type"
    echo "This option skips restoring some NVRAM variables which are unique"
    echo "  to a specific router, such as the Router Name, which by default is"
    echo "  based on the router MAC address."
    echo "If you are restoring to a different router from that used to generate the"
    echo "  nvram-save, you should answer Y to the migration restore prompt"
    #################################################################################EIC hack###############################
    printf '%b%s' "$cRESET" "Perform a migration restore [Y/N]? "
    read -r "INPUT"
    while true; do
      case "$INPUT" in
      "Y" | "y")
        opt="-migr"
        continue
        ;;
      *)
        opt=""
        ;;
      esac
    done
  fi
else
  if [ "$macsave" = "MIGR" ]; then
    echo "Performing a migration restore from a migration nvram-save"
    opt="-clean"
  else
    echo "Migration restore from a full nvram-save is not supported on this save file"
    echo ""
    printf '%s' "Do you want to proceed [Y/N]? "
    read -r "INPUT"
    while true; do
      case "$INPUT" in
      "Y" | "y")
        break
        ;;
      *)
        printf '%b\n\t%s\n\n%b' "$cBYEL" "Restore aborted" "$cRESET"
        exit 2
        ;;
      esac
    done
    opt=""
  fi
fi

#grep -q "Clean restore supported" $dwd/$restore_file
#if [ $? = 0 ]; then
if [ "$opt" = "" ]; then
  echo ""
  #################################################################################EIC hack###############################
  printf '%b' "$cBYEL"
  echo "A clean restore will restore only those NVRAM settings"
  echo "  which have been initialized by the router after a factory reset"
  echo "This option should by used when reverting to an earlier"
  echo "  firmware level, when updating past a single firmware release,"
  echo "  when moving betwwen OEM and Merlin firmware or"
  echo "  to ensure any special use or obsolete NVRAM variables are cleared"
  echo "If in doubt, answer Y to perform a clean restore"
  #################################################################################EIC hack###############################
  printf '\n%b%s' "$cRESET" "Perform a clean restore and remove unsed NVRAM variables [Y/N]? "
  read -r "INPUT"
  while true; do
    case "$INPUT" in
    "Y" | "y")
      opt="-clean"
      break
      ;;
    *)
      printf '%b\n\t%s\n\n%b' "$cBYEL" "Restore aborted" "$cRESET"
      exit 2
      #opt=""
      ;;
    esac
  done
fi
#################################################################################EIC hack###############################
start=$(date +%s)
########################################################################################################################
# Run the restore
sh "$dwd"/"$restore_file" "$opt"

# Update runlog
codelevel=$(grep "codelevel=" "$dwd"/"$restore_file" | awk -F'=' '{print $2}')
version=$(grep "#VERSION=" "$dwd"/"$restore_file" | awk -F'=' '{print $2}')

############################################################################################EIC Hack ######################################
end=$(date +%s)
difftime=$((end - start))
echo "$(printf '%-20s' "$scr_name")$(printf '%-30s' "$restore_version$space4$(date)$space4")$(printf '%-28s' "$codelevel") #Version=$version $((difftime / 60)) minutes and $((difftime % 60)) seconds elapsed" >>"$runlog"
printf '%b' "$cBGRE"
# List stats for this invocation
tail -1 "$runlog"
printf '%b' "$cRESET"
###########################################################################################################################################

echo ""
exit 0
